torch_em/self_training/fix_match.py

```diff
@@ -0,0 +1,309 @@
+        logger [TorchEmLogger] - (default: SelfTrainingTensorboardLogger)
+        momentum [float] - (default: 0.999)
+        **kwargs - keyword arguments for torch_em.DataLoader
+    """
+
+    def __init__(
+        self,
+        model,
+        unsupervised_train_loader,
+        unsupervised_loss,
+        pseudo_labeler,
+        supervised_train_loader=None,
+        unsupervised_val_loader=None,
+        supervised_val_loader=None,
+        supervised_loss=None,
+        unsupervised_loss_and_metric=None,
+        supervised_loss_and_metric=None,
+        logger=SelfTrainingTensorboardLogger,
+        source_distribution=None,
+        **kwargs
+    ):
+        # Do we have supervised data or not?
+        if supervised_train_loader is None:
+            # No. -> We use the unsupervised training logic.
+            train_loader = unsupervised_train_loader
+            self._train_epoch_impl = self._train_epoch_unsupervised
+        else:
+            # Yes. -> We use the semi-supervised training logic.
+            assert supervised_loss is not None
+            train_loader = supervised_train_loader if len(supervised_train_loader) < len(unsupervised_train_loader)\
+                else unsupervised_train_loader
+            self._train_epoch_impl = self._train_epoch_semisupervised
+
+        self.unsupervised_train_loader = unsupervised_train_loader
+        self.supervised_train_loader = supervised_train_loader
+
+        # Check that we have at least one of supvervised / unsupervised val loader.
+        assert sum((
+            supervised_val_loader is not None,
+            unsupervised_val_loader is not None,
+        )) > 0
+        self.supervised_val_loader = supervised_val_loader
+        self.unsupervised_val_loader = unsupervised_val_loader
+
+        if self.unsupervised_val_loader is None:
+            val_loader = self.supervised_val_loader
+        else:
+            val_loader = self.unsupervised_train_loader
+
+        # Check that we have at least one of supvervised / unsupervised loss and metric.
+        assert sum((
+            supervised_loss_and_metric is not None,
+            unsupervised_loss_and_metric is not None,
+        )) > 0
+        self.supervised_loss_and_metric = supervised_loss_and_metric
+        self.unsupervised_loss_and_metric = unsupervised_loss_and_metric
+
+        super().__init__(
+            model=model, train_loader=train_loader, val_loader=val_loader,
+            loss=Dummy(), metric=Dummy(), logger=logger, **kwargs
+        )
+
+        self.unsupervised_loss = unsupervised_loss
+        self.supervised_loss = supervised_loss
+
+        self.pseudo_labeler = pseudo_labeler
+
+        if source_distribution is None:
+            self.source_distribution = None
+        else:
+            self.source_distribution = torch.FloatTensor(source_distribution).to(self.device)
+
+        self._kwargs = kwargs
+
+    def get_distribution_alignment(self, pseudo_labels, label_threshold=0.5):
```

Comment:
I didn't fully go through this, and I assume you adapted it from our previous code for the distro alignment.
Still, it would be good to check this again and to give some comments / formulas here.